ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Offset.scala

```diff
@@ -3,47 +3,40 @@
 
 package com.daml.ledger.participant.state.v1
 
-import com.digitalasset.daml.lf.data.Ref.LedgerString
-import com.digitalasset.daml.lf.data.Ref
+import java.io.InputStream
+
+import com.google.protobuf.ByteString
 
 /** Offsets into streams with hierarchical addressing.
   *
   * We use these [[Offset]]'s to address changes to the participant state.
-  * We allow for array of [[Long]] to allow for hierarchical addresses.
-  * These [[Long]] values are expected to be positive. Offsets are ordered by
-  * lexicographic ordering of the array elements.
+  * Offsets are opaque values that must be must be strictly
+  * increasing according to lexicographical ordering.
   *
-  * A typical use case for [[Offset]]s would be addressing a transaction in a
-  * blockchain by `[<blockheight>, <transactionId>]`. Depending on the
-  * structure of the underlying ledger these offsets are more or less
-  * nested, which is why we use an array of [[Long]]s. The expectation is
-  * though that there usually are few elements in the array.
+  * Ledger implementations are advised to future proof their design
+  * of offsets by reserving the first (few) bytes for a version
+  * indicator, followed by the specific offset scheme for that version.
+  * This way it is possible in the future to switch to a different versioning
+  * scheme, while making sure that previously created offsets are always
+  * less than newer offsets.
   *
   */
-final case class Offset(private val xs: Array[Long]) extends Ordered[Offset] {
-  def toLedgerString: Ref.LedgerString =
-    // It is safe to concatenate number and "-" to obtain a valid transactionId
-    Ref.LedgerString.assertFromString(components.mkString("-"))
-
-  override def toString: String = toLedgerString
+final class Offset(private val value: ByteString) extends AnyVal with Ordered[Offset] {
```

Comment:
Just had a thought: shouldn't `Offset` have `equals` and `hashCode`? Sure, they're opaque, but you should at least be able to put them in a `Set` or check that a value sent over the wire is the same as one in memory.

It didn't need it before because it was a `case class` before.

Reply:
Value classes don't exist at runtime, so `ByteString`'s `hashCode` and `equals` are used already.

Reply:
Makes sense. I'd feel a lot more comfortable if `Offset` had some test cases though. (Not sure why I didn't notice that before.)

Reply:
There's not a whole lot to `Offset` itself. There are test cases for `KVOffset`, but that is around checking the bit shifting magic (which as you pointed out only worked incidentally).